lib/notifications/open.dart

```diff
@@ -0,0 +1,229 @@
+    unawaited(navigator.push(route));
+  }
+
+  NotificationNavigationData? _tryParsePayload(
+    BuildContext context,
+    Map<Object?, Object?> payload,
+  ) {
+    try {
+      return NotificationNavigationData.fromIosApnsPayload(payload);
+    } on FormatException catch (e, st) {
+      assert(debugLog('$e\n$st'));
+      final zulipLocalizations = ZulipLocalizations.of(context);
+      showErrorDialog(context: context,
+        title: zulipLocalizations.errorNotificationOpenTitle);
+      return null;
+    }
+  }
+}
+
+class NotificationNavigationData {
+  final Uri realmUrl;
+  final int userId;
+  final Narrow narrow;
+
+  NotificationNavigationData({
+    required this.realmUrl,
+    required this.userId,
+    required this.narrow,
+  }) : assert(narrow is TopicNarrow || narrow is DmNarrow);
+
+  /// Parses the iOS APNs payload and retrieves the information
+  /// required for navigation.
+  factory NotificationNavigationData.fromIosApnsPayload(Map<Object?, Object?> payload) {
+    if (payload case {
+      'zulip': {
+        'user_id': final int userId,
+        'sender_id': final int senderId,
+      } && final zulipData,
+    }) {
+      final eventType = zulipData['event'];
+      if (eventType != null && eventType != 'message') {
+        // On Android, we also receive "remove" notification messages, tagged
+        // with an `event` field with value 'remove'. As of Zulip Server 10,
+        // however, these are not yet sent to iOS devices, and we don't have a
+        // way to handle them even if they were.
+        //
+        // The messages we currently do receive, and can handle, are analogous
+        // to Android notification messages of event type 'message'. On the
+        // assumption that some future version of the Zulip server will send
+        // explicit event types in APNs messages, accept messages with that
+        // `event` value, but no other.
+        throw const FormatException();
+      }
+
+      final String realmUrl;
+      switch (zulipData) {
+        case {'realm_url': final String value}:
+          realmUrl = value;
+        case {'realm_uri': final String value}:
+          realmUrl = value;
+        default:
+          throw const FormatException();
+      }
+
+      final Narrow narrow = switch (zulipData) {
+        {
+          'recipient_type': 'stream',
+          // TODO(server-5) remove this comment.
+          // We require 'stream_id' here but that is new from Server 5.0,
+          // resulting in failure on pre-5.0 servers.
+          'stream_id': final int streamId,
+          'topic': final String topic,
+        } =>
+          TopicNarrow(streamId, TopicName(topic)),
+
+        {'recipient_type': 'private', 'pm_users': final String pmUsers} =>
+          DmNarrow(
+            allRecipientIds: pmUsers
+              .split(',')
+              .map((e) => int.parse(e, radix: 10))
+              .toList(growable: false)..sort(),
```

Comment:
nit: the sort is a significant step, so best to keep it visible and not tuck it in:
```suggestion
              .toList(growable: false)
              ..sort(),
```